expression/builtin_time.go

```diff
@@ -1770,6 +1770,52 @@ func getTimeZone(ctx context.Context) *time.Location {
 	return ret
 }
 
+// isDuration returns a boolean indicating whether the str is the format of duration.
+func isDuration(str string) bool {
+	const durationArgReg = `^(|[-]?)(|\d{1,2}\s)(\d{2,3}:\d{2}:\d{2}|\d{1,2}:\d{2}|\d{1,6})(|\.\d*)$`
+	r, _ := regexp.Compile(durationArgReg)
+	return r.MatchString(str)
+}
+
+// strDatetimeAddDuration adds duration to datetime string, returns a dutam value.
+func strDatetimeAddDuration(d string, arg1 types.Duration) (result types.Datum, err error) {
+	arg0, err := types.ParseTime(d, mysql.TypeDatetime, getFsp(d))
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	tmpDuration := arg0.Add(arg1)
+	resultDuration, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	if getFsp(d) != 0 {
+		tmpDuration.Fsp = types.MaxFsp
```

Comment:
Is this MySQL's behavior ?
By the way, you can `fsp := getFsp(d)` to avoid call `getFsp` again and again.